### ISSUE TITLE
FineTuning improvements

### ISSUE THREAD

**Issue by nevedomski** [nevedomski/mistral_ner#63]
# Test and Validate Tier 1 FineTuning Improvements

## Background

All Tier 1 improvements from NER-6 have been successfully implemented with 92% test coverage. Now we need to empirically test and measure the actual F1 score improvement.

## Implemented Improvements (from NER-6)

1. **Task 1: Focal Loss Implementation** ✅
   * Created `src/losses.py` module with FocalLoss, LabelSmoothingLoss, and ClassBalancedLoss
   * Auto-computed alpha weights based on class frequencies (83% O-tokens vs 17% entities)
   * Configurable gamma parameter (default 2.0) for focusing on hard examples
   * Factory function `create_loss_function()` for easy loss selection
2. **Task 2: LoRA Module Expansion** ✅
   * Rank increased: 16 → 32 (2x capacity)
   * Alpha increased: 32 → 64 (proportional scaling)
   * Target modules expanded: Added gate_proj, up_proj, down_proj (MLP layers)
   * Total modules: 7 (q_proj, k_proj, v_proj, o_proj + 3 MLP projections)
3. **Task 3: QLoRA Migration** ✅
   * Migrated from 8-bit to 4-bit quantization
   * Using NF4 (NormalFloat4) quantization type
   * Expected memory reduction: \~40%
   * Maintains model quality while reducing VRAM usage
4. **Task 4: Advanced Learning Rate Scheduling** ✅
   * Created `src/schedulers.py` module with multiple scheduler types
   * Implemented: linear, cosine, cosine_with_restarts, polynomial
   * Layer-wise learning rate decay support (LayerwiseLearningRateScheduler)
   * Configurable warmup steps and scheduler-specific parameters

## Objective

Empirically test the implemented improvements to measure actual F1 score gain and validate that we achieve the target F1 of 0.85+ (from baseline 0.80).

### 1\. Baseline Measurement

```bash
# Train with original config (before improvements)
python scripts/train.py --config configs/default.yaml --num-train-epochs 3
```

### 2\. Test Each Improvement Individually

```bash
# Test Focal Loss only
python scripts/train.py --loss-function focal --num-train-epochs 3

# Test LoRA expansion only
python scripts/train.py --lora-r 32 --lora-alpha 64 --num-train-epochs 3

# Test QLoRA (4-bit) only
python scripts/train.py --quantization-bits 4 --num-train-epochs 3

# Test advanced schedulers only
python scripts/train.py --lr-scheduler-type cosine_with_restarts --num-train-epochs 3
```

### 3\. Test Combined Improvements

```bash
# All Tier 1 improvements together
python scripts/train.py \
    --loss-function focal \
    --lora-r 32 \
    --lora-alpha 64 \
    --quantization-bits 4 \
    --lr-scheduler-type cosine_with_restarts \
    --num-train-epochs 5
```

### 4\. Performance Metrics to Track

* F1 score (primary metric, target: 0.85+)
* Precision and Recall
* Training time
* Memory usage (VRAM)
* Convergence speed (loss curves)

## Acceptance Criteria

- [ ] Baseline F1 score documented (\~0.80)
- [ ] Individual improvement impacts measured
- [ ] Combined improvements achieve F1 ≥ 0.85
- [ ] Memory usage reduced by \~40% with QLoRA
- [ ] Training convergence improved with focal loss
- [ ] Results documented in project README

## Next Steps

If target F1 (0.85+) not achieved, consider implementing Tier 2 improvements:

* Class-weighted loss functions
* Batch balancing strategies
* Advanced optimizer configurations
* Enhanced evaluation metrics
